mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/query/parsing/nativequery/impl/NativeQueryParser.java

```diff
@@ -57,301 +67,300 @@
  * @author Guillaume Smet
  */
 @BuildParseTree
-public class NativeQueryParser extends BaseParser<MongoDBQueryDescriptorBuilder> {
-
-	final MongoDBQueryDescriptorBuilder builder;
+public class NativeQueryParser extends BaseParser<MongoDBQueryDescriptor> {
+
+	// TODO There are many more query types than what we support.
+	protected static List<String> RESERVED = Arrays.asList(
+			"findOne", "findAndModify", "find", "insertOne", "insertMany", "insert",
+			"remove", "deleteOne", "deleteMany", "updateOne", "updateMany", "update",
+			"replaceOne", "count", "aggregate", "distinct", "mapReduce" );
+	protected MongoDBQueryDescriptorBuilder builder = new MongoDBQueryDescriptorBuilder();
+	protected CliQueryInfo cliQueryInfo = new CliQueryInfo();
+
+	private static class CliQueryInfo {
+		String operationName;
+		String collectionName;
+		List<String> pathExpression = new ArrayList<>();
+		List<String> arguments = new ArrayList<>();
+	}
+	private Action<MongoDBQueryDescriptor> processPathExpression = context -> {
+		List<String> identifiers = cliQueryInfo.pathExpression;
+		if ( identifiers.size() <= 2 ) {
+			context.getParseErrors().add(
+					createParseError(
+							context, 2, "CLI query should match 'db.<COLLECTION>.<OPERATION>'"
+							)
+					);
+			return false;
+		}
+		if ( !"db".equals( identifiers.get( 0 ) ) ) {
+			context.getParseErrors().add(
+					createParseError(
+							context, 2, "First identifier should be 'db'"
+							)
+					);
+			return false;
+		}
+		cliQueryInfo.operationName = identifiers.get( identifiers.size() - 1 );
+		cliQueryInfo.collectionName = String.join( ".", identifiers.subList( 1, identifiers.size() - 1 ) );
+		if ( RESERVED.contains( cliQueryInfo.collectionName ) ) {
+			context.getParseErrors().add(
+					createParseError(
+							context,
+							identifiers.size(),
+							"Collection name should be different from the allowed operations"
+							)
+					);
+			return false;
+		}
 
-	public NativeQueryParser() {
-		this.builder = new MongoDBQueryDescriptorBuilder();
-	}
+		return true;
+	};
+
+	private Action<MongoDBQueryDescriptor> processCliQueryInfo = context -> {
+		builder.setCollection( cliQueryInfo.collectionName );
+		switch ( cliQueryInfo.operationName ) {
+			case "find": {
+				builder.setOperation( Operation.FIND );
+				builder.setCriteria( cliQueryInfo.arguments.get( 0 ) );
+				if ( cliQueryInfo.arguments.size() > 1 ) {
+					builder.setProjection( cliQueryInfo.arguments.get( 1 ) );
+				}
+				break;
+			}
```

Comment:
What annoys me a bit with this new approach is that we are far less selective into what would be a valid option for `find()` (same for any other option). We treat all arguments equal and I'm wondering if it's such a good idea.

I think I would have chosen to keep the existing approach and have a default catch all option that would:
 * say that the syntax is incorrect if the action is in the list of supported command
 * either say the command is not supported for now (for existing commands that we don't support yet),
 * or is an invalid command.

i'm not sure though how it would be compatible with saying that the syntax of `find()` is not valid with an error in the parser. That's probably why I thought we would do that in the query builder instead.

Not sure what's the best outcome here. @DMasherov WDYT about it?